### ISSUE TITLE
Apply pagination wherever required

### ISSUE THREAD

**Issue by Mayank0255** [Mayank0255/StackOverflowClone#2]
#### Back-end

Create a middleware for pagination

_Endpoints which are to be changed:_
* `/api/users`
* `/api/posts`
* `/api/posts/top`
* `/api/tags`
* `/api/posts/tag/:tagname`
* `/api/posts/answers/:id`
* `/api/posts/answers/:id`

#### Front-end

Add pagination in the front-end

**Comment by bishalr0y** [comment]
Hey, I would like to work on this issue.  Can you guide me on how to get started with it?

**Comment by devanshiv123** [comment]
@Mayank0255 Can you assign this issue to me?

**Comment by Mayank0255** [comment]
@devanshiv123 sure, go ahead

**Comment by devanshiv123** [comment]
@Mayank0255 I am getting the the following error while setting up project(after npm run dev)
![image](https://user-images.githubusercontent.com/43826157/131402652-481e3946-bbac-48a2-9ec4-ec8b0808be4b.png)

**Comment by Mayank0255** [comment]
@devanshiv123 
Try removing all the question marks from the shown areas and then try again and let me know

**Comment by devanshiv123** [comment]
I have removed the question marks and now I am getting another error:
![image](https://user-images.githubusercontent.com/43826157/131448665-4377a7b0-9d3e-40b0-9339-e62a9fb915e3.png)

**Comment by Mayank0255** [comment]
@devanshiv123 
Can you go to the sidebar file in whoch it is exported and check whether it's exported correctly or not?

**Comment by devanshiv123** [comment]
@Mayank0255 Yes I think its exported correctly

**Comment by Mayank0255** [comment]
@devanshiv123 
It's a new error actually, I haven't gotten this ever.

Can you look into it if possible? I will also try to figure it out.

Btw I am assuming that you haven't changed anything in the code right now?

**Comment by devanshiv123** [comment]
@Mayank0255 No I havn't made any changes I have just made the setup till now. I will try to figure out, maybe an issue with dependency.

**Comment by Mayank0255** [comment]
@devanshiv123 yes, correct, it's been sometime time there had been any changes so might be some functions or methods may have depreciated with the new versions

**Comment by Mayank0255** [comment]
@devanshiv123 any updates?

**Comment by devanshiv123** [comment]
@Mayank0255 I am still getting those errors. I have tried for a day but maybe something is wrong. For now you can unassign me from the issue, I will first try to setup correctly and then take up any other issue.

**Comment by Mayank0255** [comment]
@devanshiv123 Oh, ok

Any other reason that you wanna work on some other issue?

**Comment by devanshiv123** [comment]
> @devanshiv123 Oh, ok
> 
> Any other reason that you wanna work on some other issue?

@Mayank0255 No nothing as such, I will surely try to contribute if I get the setup properly after resolving errors.

**Comment by Mayank0255** [comment]
@devanshiv123 Cool

**Comment by alceil** [comment]
Is anyone working on this issue? Can i work on this issue?

**Comment by Mayank0255** [comment]
@alceil Yes you can work on this issue,

I noticed that you are willing to work on two more issues, which are also available but I would recommend doing one first and then going on others, so that I can just assign you one at the moment and the other two are open for other people as well in the meantime.

You can choose which one do you want to start with? I would recommend this one as this one is the smallest one among the three and this will help you understand the codebase faster as well

**Comment by kunalkumar007** [comment]
Hy @alceil @Mayank0255, if @alceil decides to work on another issue, then I am open to this one. Let me know further. 😃

**Comment by Mayank0255** [comment]
@alceil can you please clarify that from which issue are you willing to begin?

@kunalkumar007 kindly give the person 1 day to get back to us, if he decides to take up this issue then I can recommend you to take up #31 this issue, which might be something new and for which a guide has also been referred. If he decides to take up anyother issue then I would love you to be part of this issue.

If @alceil doesn't gives us an update by the EOD then I can happily assign this to @kunalkumar007 as I am looking for some active participants as well :-)

**Comment by JigmeT123** [comment]
Is this still active ? Can I work on it

**Comment by Mayank0255** [comment]
> Is this still active ? Can I work on it

Yes it is active, @kunalkumar007 is also willing to work on this, so @kunalkumar007 would you like to work with @JigmeT123 on this or do you want to work on this separately?

**Comment by kunalkumar007** [comment]
As more developers are interested to work on this issue, I would try another issue.

**Comment by Mayank0255** [comment]
@kunalkumar007 sure, kindly comment on that particular issue so that I can assign you that.

**Comment by Mayank0255** [comment]
@JigmeT123 I am then assigning this issue to you, is that ok?

**Comment by JigmeT123** [comment]
@Mayank0255 sure I will take this

**Comment by Mayank0255** [comment]
@JigmeT123 Cool. Looking forward to a PR soon!!

**Comment by luko0610** [comment]
@Mayank0255 In case this doesn't make any progress I'd be happy to take over.

**Comment by Mayank0255** [comment]
@luko0610 Okay so @JigmeT123 two days has completed since I asked him about the updates.

Should I assign it to you then? And may I ask that approximately after how much time can I expect a PR?

**Comment by luko0610** [comment]
@Mayank0255 I'm currently at it. Frontend pagination component is nearly done (still needs some style improvements).
For now pagination is client side. If you want I can also make a PR for the backend to add server side pagination.

**Comment by Mayank0255** [comment]
> @Mayank0255 I'm currently at it. Frontend pagination component is nearly done (still needs some style improvements). For now pagination is client side. If you want I can also make a PR for the backend to add server side pagination.

Great progress @luko0610 ,

Can you elaborate on the part of server side pagination? If pagination has been setup on the frontend then what will you be doing in the backend?

**Comment by luko0610** [comment]
Of course. Basically if we only set it up on the frontend, the application would always receive all entries from the backend and then handle pagination on the client side. That's okay for a small number of entries, but can become an issue, when having thousands of entries.

A solution for this issue would be to do the pagination server side. That means when fetching the entries from the backend we always send the currentPage and pageSize we want returned, and then only get sent back entries we want to currently show in the frontend.

I would for now implement it client side, but I could of course make another PR to add server side pagination later on.

**Comment by Mayank0255** [comment]
> @Mayank0255 Of course. Basically if we only set it up on the frontend, the application would always receive all entries from the backend and then handle pagination on the client side. That's okay for a small number of entries, but can become an issue, when having thousands of entries.
> 
> A solution for this issue would be to do the pagination server side. That means when fetching the entries from the backend we always send the currentPage and pageSize we want returned, and then only get sent back entries we want to currently show in the frontend.
> 
> I would for now implement it client side, but I could of course make another PR to add server side pagination later on.

@luko0610 I think it's a good idea, I have two things to ask about it,
1. After implementing that in the `API`, will you be calling the endpoint with updated page whenever next page or a specific page button is pressed?

2. There's one important thing to keep in mind while adding pagination, which is get access to `last page number (can be names as lastPageNumber)` at all the time as in the client side it is necessary to show the page numbers from the first to the last page.

I would request you to make a an issue related to this on the backend repo, and then add a brief description related to it so that I can assign you that and the project management runs smoothly.
